src/fondant/executor.py

```diff
@@ -189,43 +191,109 @@ def _write_data(self, dataframe: dd.DataFrame, *, manifest: Manifest):
+                )
+
+                # Get the most recent file based on the file creation date time
+                manifest_file = max(matching_manifests, key=self.filesystem.created)
+            else:
+                manifest_file = matching_manifests[0]
+
+            return Manifest.from_file(manifest_file)
+
+        logger.info("No matching execution for component detected")
 
-    def _has_matching_execution(self) -> bool:
-        """Function that checks if there is an existing previous matching execution."""
-        # TODO: implement
-        return False
+        return None
+
+    def _is_previous_cached(self, input_manifest: Manifest) -> bool:
+        """
+        Checks whether the previous component's output is cached based on its run ID.
+
+        This function compares the run ID of the input manifest
+         (representing the previous component) with the run ID of the current component metadata.
+        If the run IDs are different, it indicates that the previous component's output belongs to
+        another pipeline run, implying that it is cached. Otherwise, if the run IDs match, it
+        suggests that the previous component was not cached and had to execute to produce the
+         current output.
+
+        Args:
+            input_manifest: The manifest representing the output of the previous component.
+
+        Returns:
+            True if the previous component's output is cached, False otherwise.
+        """
+        previous_component_id = input_manifest.component_id
+
+        if input_manifest.run_id == self.metadata.run_id:
+            logger.info(
+                f"Previous component `{previous_component_id}` is not cached."
+                f" Invalidating cache for current and subsequent components",
+            )
+            return False
+
+        logger.info(
+            f"Previous component `{previous_component_id}` run was cached. "
+            f"Cached pipeline id: {input_manifest.run_id}",
+        )
+        return True
 
     def execute(self, component_cls: t.Type[Component]) -> None:
-        """Execute a component.
+        """
+        Execute a component.
 
         Args:
             component_cls: The class of the component to execute
         """
-        matching_execution_exists = self._has_matching_execution()
 
-        if matching_execution_exists:
-            logger.info("Previous matching execution found")
-        else:
-            logger.info("No previous matching execution found")
+        def _run_execution(inner_input_manifest: Manifest) -> Manifest:
+            logging.info("Executing component")
+            component = component_cls(self.spec, **self.user_arguments)
+            output_df = self._execute_component(
+                component,
+                manifest=inner_input_manifest,
+            )
+            inner_output_manifest = input_manifest.evolve(component_spec=self.spec)
```

Comment:
```suggestion
            inner_output_manifest = inner_input_manifest.evolve(component_spec=self.spec)
```

Maybe it's better to move this into a separate method to prevent these kind of issues.